x-pack/plugin/sql/jdbc/src/test/java/org/elasticsearch/xpack/sql/jdbc/JdbcDatabaseMetaDataTests.java

```diff
@@ -8,13 +8,50 @@
 
 import org.elasticsearch.test.ESTestCase;
 
+import static org.elasticsearch.xpack.sql.jdbc.JdbcConfiguration.EscapeWildcard.ALWAYS;
+import static org.elasticsearch.xpack.sql.jdbc.JdbcConfiguration.EscapeWildcard.AUTO;
+import static org.elasticsearch.xpack.sql.jdbc.JdbcConfiguration.EscapeWildcard.NEVER;
+import static org.elasticsearch.xpack.sql.jdbc.JdbcDatabaseMetaData.escapeStringPattern;
+
 public class JdbcDatabaseMetaDataTests extends ESTestCase {
 
+    final char ESC = '\\';
     private JdbcDatabaseMetaData md = new JdbcDatabaseMetaData(null);
 
     public void testSeparators() throws Exception {
         assertEquals(":", md.getCatalogSeparator());
         assertEquals("\"", md.getIdentifierQuoteString());
         assertEquals("\\", md.getSearchStringEscape());
     }
+
+    public void testEscapeWildcardNever() {
+        assertEquals("foo", escapeStringPattern("foo", NEVER));
+        assertEquals("f__", escapeStringPattern("f__", NEVER));
+        assertEquals("f__b", escapeStringPattern("f__b", NEVER));
+        assertEquals("f%_b", escapeStringPattern("f%_b", NEVER));
+        assertEquals("f\\_b", escapeStringPattern("f\\_b", NEVER));
+        assertEquals(null, escapeStringPattern(null, NEVER));
+    }
+
+    public void testEscapeWildcardAlways() {
+        assertEquals("foo", escapeStringPattern("foo", ALWAYS));
+        assertEquals("%", escapeStringPattern("%", ALWAYS));
+        assertEquals("\\%a", escapeStringPattern("%a", ALWAYS));
+        assertEquals("f\\_", escapeStringPattern("f_", ALWAYS));
+        assertEquals("f\\\\_", escapeStringPattern("f\\_", ALWAYS));
+        assertEquals("f\\%\\_", escapeStringPattern("f%_", ALWAYS));
+    }
+
+    public void testEscapeWildcardAuto() {
+        assertEquals("foo", escapeStringPattern("foo", AUTO));
```

Comment:
@bpintea please double check if this is what you would expect. Essentially AUTO tries to first determine if there's any escaping - if it, it will escape `_` but not `%` and the reason behind that is a bit of heuristics - % seems to be used rarely while `_` is more common in names.
There's also the change of using `%` inside a name however typically there would be some letters thus somewhat reducing the clash possibility. Also I would image there are higher chances of using a pattern such as `%bar` vs actually `\%bar`.

Reply:
First, this parameter maps relatively well over what ODBC's [`SQL_ATTR_METADATA_ID`](https://docs.microsoft.com/en-us/sql/odbc/reference/develop-app/pattern-value-arguments) does: if not set (/`false`), the catalog functions need to accept search patterns (i.e. `foo_ar` matching `foobar`) in arguments and if set (/`true`), the arguments must be treated as literals (i.e. `foo_ar` will not match `foobar` or anything else than `foo_ar`).

Then, expectancy-wise, I think this can be made to work.
The default for `SQL_ATTR_METADATA_ID` if `false` (i.e. `_` and `%` need to work as patterns).\
The experience so far shows that no "common" application changes the default value **and** there aren't obvious / "quick" controls available to the user to turn this on. Which means that in the cases where the attribute's value isn't the default (`false`), the application has a clear intention ("i want literal matching, not pattern matching"). Which would lead to the following mapping of `SQL_ATTR_METADATA_ID` -> `compat.meta.escape.wildcards`:
* `true` (literal) -> `ALWAYS`;
* `false` (pattern) -> user configuration (new setting to be added to ODBC connection string):
  * `AUTO` (the corresponding new setting's default): the server would escape `_` if not already done (and leave `%` as is);
  * `NEVER`: if `_` is escaped, no matching needs to happen; if it isn't escaped, pattern matching must apply.

Reply:
To clarify: it's the client that does the escaping, not the server - in fact, everything including the setting, etc.. happens on the client.

Reply:
Ah, I see, sorry for the misunderstanding.
I've just run a quick test with an alias containing `%` and two of the applications that don't escape the `_` (Tableau, Qlik): they don't escape the `%` either.

So I would consider the equivalent setting in ODBC, but with only `on`/`off` values (equivalents to `ALWAYS` and `NEVER` respectively); furthermore this would apply to only pattern matching case, since the literal matching one (`SQL_ATTR_METADATA_ID` == `true`) would need to always escape.
Mentioning this just in case this might be something to consider for JDBC as well.

Reply:
The issue with ALWAYS is that it will escape `%` as well for example, `%` by itself should not be escaped but `foo%bar` might -whether that's the case or not depends on whether somebody sets ALWAYS or AUTO (which escaped `_` only).

Reply:
> But foo%bar + AUTO would still result in a pattern matching foo%bar (and not foo\%bar) right, which would still not restrict the over-matching for misbehaving apps?

No, AUTO would change that to (`foo\%bar`) - see the tests below:
```java
        assertEquals("\\%a", escapeStringPattern("%a", AUTO));
        assertEquals("f\\_", escapeStringPattern("f_", AUTO));
        assertEquals("f\\_", escapeStringPattern("f\\_", AUTO));
        assertEquals("f\\%\\_", escapeStringPattern("f%_", AUTO));
        assertEquals("f%\\_", escapeStringPattern("f%\\_", AUTO));
```

Essentially AUTO returns the patterns as they are if it detects an escaping pattern otherwise it will escape non-escaped or invalid patterns.

Also if you check, single `%` and `_` are never escaped regardless of the setting value.
Again, take a look at the tests and see whether there's a case that it's not tested and I'll add it in for clarification.